### PR TITLE
use options keeps users from choosing between Validate/ValidateCustom and Generate/GenerateCustom

### PR DESCRIPTION
**Pull request by d-fal**
I've been using this package for quite a while and it has helped me thousands of times. However, sometimes I get confused in using custom validators/generators. I think we can make it better if you agree on using options in main functions, say, Validate and Generate funcs.

### CODE REVIEW

File: totp/totp.go [hunk 1]
```diff
@@ -74,13 +71,55 @@ type ValidateOpts struct {
...
+	}
+}
+
 // GenerateCodeCustom takes a timepoint and produces a passcode using a
 // secret and the provided opts. (Under the hood, this is making an adapted
 // call to hotp.GenerateCodeCustom)
-func GenerateCodeCustom(secret string, t time.Time, opts ValidateOpts) (passcode string, err error) {
-	if opts.Period == 0 {
-		opts.Period = 30
+func GenerateCodeCustom(secret string, t time.Time, validateOpts ...ValidateOpt) (passcode string, err error) {
```
Comment (pquerna) [inline totp/totp.go:114]: this breaks semver compat right?

can we introduce new functions so v1 remains API compataible?

Comment (d-fal) [reply totp/totp.go:114]: that's right, it breaks semver. We can introduce new funcs and add **deprecated** tags to the current ones, if possible.

Comment (pquerna) [reply totp/totp.go:114]: SGTM

Comment (d-fal) [reply totp/totp.go:114]: what's the next move? should I make any changes?

Comment (pquerna) [reply totp/totp.go:114]: I think the path forward is to make a new function, GenerateCodeCustomOpts, and label GenerateCodeCustom as depreciated, See GenerateCodeCustomOpts.

I think Generate, and ValidateCustom needs the same treatment.

I don't want to break ABI compatibility for this kind of change - its a good thing to do, but on the aesthetics side, so I think new funcs that don't break the API are the way to go.

Comment (d-fal) [reply totp/totp.go:114]: I'll work on it...

**Comment by d-fal** [comment]
voila, as requested.

File: examples/custom_totp_generation/main.go [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package main
+
+/*
+ In this example, we want to show the definition of custom-generated totp
+*/
+func main() {
+
```
Comment (pquerna) [inline examples/custom_totp_generation/main.go:7]: looks like no example here... so remove the file for now :) ?

File: hotp/hotp.go [hunk 3]
```diff
@@ -18,9 +18,10 @@
 package hotp
 
 import (
-	"github.com/pquerna/otp"
 	"io"
 
+	"github.com/pquerna/otp"
```
Comment (pquerna) [inline hotp/hotp.go:23]: double import of the same package?

File: examples/custom_totp_generation/main.go (hunk 2 above)
Comment (d-fal) [reply examples/custom_totp_generation/main.go:7]: :) yes please, I'll add it soon.

**Comment by d-fal** [comment]
sorry for the late update. I tried to fix the errors and added examples and tests.

**Review by d-fal — COMMENTED** [review]
the errors were resolved and tests were successful at my end.

**Review by d-fal — COMMENTED** [review]
I reviewed the code and the conflicts that were made by my IDE were resolved. I don't see any conflicts at my fork: https://github.com/d-fal/otp

**Comment by d-fal** [comment]
there were no updates by maintainers.

**Pull request closed without merging**